<a target="_blank" href="https://colab.research.google.com/github/cesarschoollectures/am-labs/blob/main/assignments/E01_Decision_Tree.ipynb">
<img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# Aprendizado de Máquina

Nesta atividade, você irá trabalhar com o dataset MNIST handwritten digits utilizando modelos de classificação do sklearn.

O foco NÃO é apenas obter bons resultados, mas garantir que o experimento seja:
- correto
- reprodutível
- bem estruturado
- criticamente analisado

# Dicas importantes

## Sobre o dataset (MNIST handwritten digits)

- Utilize os arquivos `*.csv` disponibilizados via Google Classroom
- Use: `as_frame=False`
- Use: `mnist_784`

# Questão 1

Implemente uma função load_data(seed) que:

Carregue o dataset `MNIST handwritten digits` (arquivos de treino e teste)
Realize a separação do conjunto de treino como treino e validação
Utilize `train_test_split` com controle de aleatoriedade (seed)
Retorne: `X_train`, `X_val`, `y_train`, `y_val`

Depois responda: 
É necessário normalizar os dados para esse tipo de modelo? Justifique.

**Solução**:

In [ ]:
# TODO: implemente load_data

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv("mnist_train.csv")

In [4]:
def load_data(seed):
    X = df.drop("label", axis=1)
    y = df["label"]

    X_train, X_val, y_train, y_val = train_test_split(
        X,
        y,
        test_size=0.2,
        random_state=seed,
        stratify=y
    )

    return X_train, X_val, y_train, y_val

In [5]:
X_train, X_val, y_train, y_val = load_data(42)

print(X_train.shape)
print(X_val.shape)

(48000, 784)
(12000, 784)


In [6]:
X_train = X_train / 255.0
X_val = X_val / 255.0

In [7]:
X_train.head()

,1x1,1x2,1x3,1x4,1x5,1x6,1x7,1x8,1x9,1x10,...,28x19,28x20,28x21,28x22,28x23,28x24,28x25,28x26,28x27,28x28
13259,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1201,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
40980,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
24007,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
54618,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [8]:
X_train.min().min(), X_train.max().max()

(np.float64(0.0), np.float64(1.0))

Sim, é necessário normalizar os dados. O dataset MNIST possui valores de pixels entre 0 e 255. Essa diferença de escala pode prejudicar o desempenho de algoritmos de machine learning. A normalização coloca os valores na faixa de 0 a 1, o que melhora a estabilidade numérica, acelera o treinamento e melhora o desempenho dos modelos.

# Questão 2

Implemente as funções:

`train_random_forest(X_train, y_train, seed)`

## Requisitos:

Utilizar os modelos do `sklearn`
Garantir reprodutibilidade com `random_state`

**Solução**:

In [ ]:
# TODO: implemente load_data

In [9]:
from sklearn.ensemble import RandomForestClassifier

In [10]:
def train_random_forest(X_train, y_train, seed):

    model = RandomForestClassifier(
        n_estimators=100,
        random_state=seed,
        n_jobs=-1
    )

    model.fit(X_train, y_train)

    return model

In [11]:
model = train_random_forest(X_train, y_train, 42)

print(model)

RandomForestClassifier(n_jobs=-1, random_state=42)


In [12]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_val)
accuracy_score(y_val, y_pred)

0.9665

O modelo Random Forest obteve uma acurácia de 96,65% na validação. Esse resultado indica um bom desempenho do modelo na classificação dos dígitos manuscritos do dataset MNIST. A performance pode ser atribuída à capacidade do Random Forest de combinar múltiplas árvores de decisão, reduzindo overfitting e aumentando a generalização.

# Questão 3

Implemente a função:

- `evaluate(model, X_test, y_test)`

Ela deve:
- Realizar predições
- Retornar a acurácia do modelo

**Solução**:

In [ ]:
# TODO: implemente load_data

In [13]:
from sklearn.metrics import accuracy_score

In [14]:
def evaluate(model, X_test, y_test):

    y_pred = model.predict(X_test)

    acc = accuracy_score(y_test, y_pred)

    return acc

In [15]:
acc = evaluate(model, X_val, y_val)

print(acc)

0.9665


**Adicione seu texto de solução aqui**.

O valor obtido pela função evaluate é igual ao cálculo anterior porque ambos utilizam o mesmo modelo treinado e o mesmo conjunto de validação. A métrica de acurácia é determinística nesse caso, resultando no mesmo valor de desempenho.

# Questão 4

Implemente a função:

- `run_pipeline(model_type="rf", seed=42)`

Ela deve:
- Carregar os dados
- Treinar o modelo escolhido (`rf`)
- Avaliar o modelo
- Retornar a acurácia

**Solução**:

In [ ]:
# TODO: implemente load_data

In [16]:
from sklearn.ensemble import RandomForestClassifier

In [17]:
def run_pipeline(model_type="rf", seed=42):

    X_train, X_val, y_train, y_val = load_data(seed)


    if model_type == "rf":
        model = RandomForestClassifier(
            n_estimators=100,
            random_state=seed,
            n_jobs=-1
        )
    else:
        raise ValueError("Modelo não suportado")

    
    model.fit(X_train, y_train)

    
    y_pred = model.predict(X_val)
    acc = accuracy_score(y_val, y_pred)

    return acc

In [18]:
run_pipeline("rf", 42)

0.9665

O valor de acurácia permaneceu consistente devido à reprodutibilidade do experimento. O uso do parâmetro random_state garante que a divisão dos dados e o treinamento do modelo sejam determinísticos, produzindo o mesmo resultado para a mesma configuração.

In [19]:
run_pipeline("rf", seed=7)

0.9689166666666666

A pequena variação na acurácia ocorre devido às diferentes divisões do conjunto de dados durante o treino e validação. Mesmo com o uso de random_state, pequenas mudanças na distribuição dos dados podem impactar o desempenho do modelo, especialmente em algoritmos baseados em ensembles como o Random Forest.

**Em qual profundidade começa o overfitting?**
**Por que a árvore consegue 100% no treino quando max_depth=None?**

O overfitting em árvores de decisão tende a começar quando a profundidade do modelo se torna muito alta, geralmente a partir de valores elevados (por exemplo, acima de 10–20 níveis), dependendo da complexidade dos dados. Quando max_depth=None, a árvore não possui restrição de profundidade e cresce até que todas as folhas sejam puras ou contenham o número mínimo de amostras. Isso faz com que o modelo memorize completamente o conjunto de treino, resultando em 100% de acurácia no treino. No entanto, isso prejudica a capacidade de generalização, caracterizando overfitting. 

# Questão 5

Execute o pipeline para ambos os modelos:

- Random Forest

## Apresente:
- Acurácia, Precisão, Recall e F1-Score para o modelo

**Solução**:

In [ ]:
# TODO: implemente load_data

In [20]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [21]:
X_train, X_val, y_train, y_val = load_data(seed=42)

model = train_random_forest(X_train, y_train, seed=42)

y_pred = model.predict(X_val)

In [22]:
acc = accuracy_score(y_val, y_pred)
precision = precision_score(y_val, y_pred, average="weighted")
recall = recall_score(y_val, y_pred, average="weighted")
f1 = f1_score(y_val, y_pred, average="weighted")

acc, precision, recall, f1

(0.9665, 0.9665020793737443, 0.9665, 0.9664770001986567)

Os resultados obtidos pelo modelo Random Forest foram:

- Acurácia: 0.9665
- Precisão: 0.9665
- Recall: 0.9665
- F1-score: 0.9664

As métricas apresentam valores muito próximos, o que indica um desempenho equilibrado do modelo na tarefa de classificação dos dígitos manuscritos. Isso sugere que o modelo não está enviesado para nenhuma classe específica e possui boa capacidade de generalização no conjunto de validação.

# Questão 6

Execute o pipeline utilizando diferentes seeds (ex: 42 e 7).

## Analise:
- Os resultados mudaram?

## Responda:
- O experimento é reprodutível? Justifique.

**Solução**:

In [ ]:
# TODO: implemente load_data

In [23]:
run_pipeline("rf", seed=42)

0.9665

In [24]:
run_pipeline("rf", seed=7)

0.9689166666666666

Os resultados apresentam pequenas variações ao utilizar diferentes seeds (42 e 7), o que é esperado devido à natureza aleatória da divisão dos dados e do treinamento do Random Forest. No entanto, o experimento é reprodutível, pois ao fixar um valor de random_state, é possível obter sempre os mesmos resultados para aquela configuração específica. Isso garante consistência e controle sobre a aleatoriedade do processo.

# Questão 7

Para pelo menos um dos modelos:

- Compare a acurácia em treino e teste

## Responda:
- Existe overfitting?
- Qual modelo tende a sofrer mais com isso?

In [ ]:
# TODO: implemente load_data

In [25]:
X_train, X_val, y_train, y_val = load_data(seed=42)

model = train_random_forest(X_train, y_train, seed=42)

In [26]:
y_train_pred = model.predict(X_train)
train_acc = accuracy_score(y_train, y_train_pred)

In [27]:
y_val_pred = model.predict(X_val)
val_acc = accuracy_score(y_val, y_val_pred)

train_acc, val_acc

(1.0, 0.9665)

A acurácia no conjunto de treino foi de 1.0, enquanto no conjunto de teste foi de 0.9665.

Essa diferença indica a presença de leve overfitting, pois o modelo apresenta desempenho perfeito no treino, mas uma redução no desempenho ao generalizar para dados não vistos.

Modelos baseados em árvores de decisão, especialmente árvores profundas sem restrição de profundidade, são mais propensos ao overfitting. O Random Forest reduz esse problema ao combinar múltiplas árvores, mas ainda pode apresentar pequenas diferenças entre treino e teste devido à sua flexibilidade.

# Questão 8

Varie pelo menos um hiperparâmetro em cada modelo:

- Random Forest: `n_estimators`

## Analise:
- O desempenho muda significativamente?

In [ ]:
# TODO: implemente load_data

In [28]:
model_low = RandomForestClassifier(
    n_estimators=10,
    random_state=42,
    n_jobs=-1
)

model_low.fit(X_train, y_train)
y_pred_low = model_low.predict(X_val)

acc_low = accuracy_score(y_val, y_pred_low)
acc_low

0.9459166666666666

In [29]:
model_mid = RandomForestClassifier(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

model_mid.fit(X_train, y_train)
y_pred_mid = model_mid.predict(X_val)

acc_mid = accuracy_score(y_val, y_pred_mid)
acc_mid

0.9665

In [30]:
model_high = RandomForestClassifier(
    n_estimators=300,
    random_state=42,
    n_jobs=-1
)

model_high.fit(X_train, y_train)
y_pred_high = model_high.predict(X_val)

acc_high = accuracy_score(y_val, y_pred_high)
acc_high

0.968

Ao variar o número de estimadores (n_estimators) no Random Forest, observa-se que o desempenho melhora conforme o número de árvores aumenta. Com 10 estimadores, o modelo apresenta menor desempenho (0.9459), indicando menor estabilidade. Ao aumentar para 100 estimadores, há um ganho significativo (0.9665), mostrando maior robustez. Com 300 estimadores, o desempenho melhora levemente (0.968), indicando que o modelo atinge um ponto de saturação. Isso mostra que aumentar o número de árvores melhora o desempenho até certo limite, após o qual os ganhos tornam-se marginais, enquanto o custo computacional aumenta.

# Questão 9

Responda (máx. 2 parágrafos por item):

1. A acurácia é suficiente para avaliar os modelos?
2. Como você garante que o resultado não ocorreu por acaso?
3. Cite dois possíveis problemas metodológicos neste experimento.
4. O pipeline implementado é confiável? Justifique.

In [ ]:
# TODO: implemente load_data

A acurácia é uma métrica importante e bastante intuitiva, pois indica a proporção geral de acertos do modelo. No contexto do MNIST, ela já fornece uma boa visão inicial do desempenho, especialmente porque o problema é relativamente balanceado e bem definido. No entanto, a acurácia sozinha não é suficiente para uma avaliação completa. Ela não captura possíveis erros por classe nem revela desequilíbrios no comportamento do modelo. Por isso, métricas como precisão, recall e F1-score são necessárias para uma análise mais robusta e confiável.

Para garantir que os resultados não ocorreram por acaso, é fundamental utilizar estratégias de reprodutibilidade e validação. O uso do parâmetro random_state assegura que as divisões de dados e o treinamento sejam consistentes entre execuções, permitindo replicação dos experimentos. Além disso, a utilização de um conjunto de validação separado e a análise de diferentes seeds ajudam a verificar a estabilidade do modelo. Caso os resultados permaneçam consistentes em diferentes execuções, isso aumenta a confiança de que o desempenho observado não foi fruto do acaso.

Um possível problema metodológico é a ausência de validação cruzada (cross-validation), que poderia fornecer uma estimativa mais robusta do desempenho do modelo ao invés de depender de uma única divisão treino/validação. Outro problema é a possibilidade de overfitting, especialmente em modelos baseados em árvores com alta profundidade, como observado quando max_depth=None. Isso pode levar o modelo a memorizar o conjunto de treino, reduzindo sua capacidade de generalização.

O pipeline implementado é confiável do ponto de vista técnico, pois segue boas práticas como separação entre treino e validação, uso de reprodutibilidade com random_state, normalização dos dados e avaliação com múltiplas métricas.

No entanto, sua confiabilidade poderia ser ainda maior com a inclusão de validação cruzada, testes em diferentes particionamentos dos dados e comparação com outros modelos. Dessa forma, embora o pipeline seja sólido e consistente para fins educacionais e experimentais, ainda há espaço para torná-lo mais robusto em um cenário de produção.